src/app/app_main.c

```diff
@@ -0,0 +1,19 @@
+#include <stdint.h>
+#include "app_header.h"
+
+void _start(void);
+
+int main() {
+    while (1);
+    return 0;
+}
+
+__attribute__((section(".app_header"))) const app_header_t app_header = {
```

Comment:
Is there any reason as to why you've decided to create a new, non-standard section called `.app_header`? You can just use `.rodata.app_header` since you're compiling with `-fdata-sections`.

Other thing, I think you should make an `app_header.c` so that it's clear where the app header resides without necessarily adding extra stuff into `main.c`.